design/thrift-proxy-design.md

```diff
@@ -0,0 +1,215 @@
+        headers:
+          - name: <some_header>
+            exact: <value>
+      services:
+        - name: <cluster_name>
+          port: <port_number>
+```
+
+### Example
+```yaml
+apiVersion: projectcontour.io/v1
+kind: ThriftProxy
+metadata:
+  name: placeorder
+  namespace: placeorder-ns
+spec:
+  routes:
+    - condition:
+        methodName: "placeOrder"
+        headers:
+          - name: "Authorization"
+            exact: "Bearer token"
+      services:
+        - name: "order-service"
+          port: 9090
+
+```
+
+## Thrift Listener
+When a new ThriftProxy resource is discovered by Contour, a corresponding Thrift Listener is automatically added to the Envoy configuration. 
+This listener is configured with default settings to ensure seamless integration into Contour.
+### Default Settings:
+* Name: The Thrift Listener is named "ingress_thrift" for identification within the Envoy configuration.
+* Protocol: The protocol is set to "thrift" to differentiate Thrift traffic from other protocols.
+* Address and Port: The address and port are configurable, allowing users to specify the desired listening address and port number, similar to how the HTTP listener operates (default: :9090)
+* RouteConfigName: The Thrift Listener is associated with a default Route Configuration named "ingress_thrift.".
+
+The Thrift Listener includes a Thrift filter with essential configurations:
+```
+&envoy_listener_v3.Filter{
+    StatPrefix: name,
+    AccessLog:  accesslogger,
+    Trds: &envoy_thrift_v3.Trds{
+        RouteConfigName: name,
+        ConfigSource: &envoy_core_v3.ConfigSource{
+            ResourceApiVersion: envoy_core_v3.ApiVersion_V3,
+            ConfigSourceSpecifier: &envoy_core_v3.ConfigSource_Ads{
+                Ads: &envoy_core_v3.AggregatedConfigSource{},
+            },
+        },
+    },
+}
+```
+
+## Thrift RouteConfiguration
+A Thrift RouteConfiguration is essentially an organized collection of Thrift routes.
+Thrift routes are defined as follows:
+```
+&envoy_thrift_v3.Route{
+    Match: ThriftRouteMatch(rule),
+    Route: ThriftRouteAction(rule),
+}
+```
+* ThriftRouteMatch: Defines the matching rules that incoming Thrift traffic must satisfy to be routed according to this rule. These rules can be based on factors such as the Thrift method name and custom headers.
+* ThriftRouteAction: Specifies the cluster that should handle the Thrift traffic that matches the given rule. This cluster configuration includes details such as load balancing policies, fail-over strategies, and rate limiting.
+
+## Aggregated Service Discovery
+Aggregated Service Discovery (ASD) is a dynamic resource management mechanism that allows Envoy to fetch configuration resources from multiple sources, 
+aggregate them, and ensure that they are consistently and efficiently distributed to proxy instances. In the context of our project, ASD serves as the backbone of Thrift route discovery.
+
+### Enabling Aggregated Service Discovery
```

Comment:
We've wanted to take advantage of go-control-plane's xDS server implementation (to utilize ADS etc.) for quite a while, but haven't had the time to iron out details relating to making it the default. I would prefer to pursue that avenue to enable ADS rather than adding to the existing contour-specific implementation of the xDS server which we intended to get rid of at some point if ADS is required to make this work.

Reply:
Just to make sure I understand correctly what we are talking about here. 
We are talking about this snippet of code, right?
```
	switch x.config.Type {
	case contour_api_v1alpha1.EnvoyServerType:
		v3cache := contour_xds_v3.NewSnapshotCache(false, log)
		x.snapshotHandler.AddSnapshotter(v3cache)
		contour_xds_v3.RegisterServer(envoy_server_v3.NewServer(ctx, v3cache, contour_xds_v3.NewRequestLoggingCallbacks(log)), grpcServer)
	case contour_api_v1alpha1.ContourServerType:
		contour_xds_v3.RegisterServer(contour_xds_v3.NewContourServer(log, xdscache.ResourcesOf(x.resources)...), grpcServer)
	default:
		// This can't happen due to config validation.
		log.Fatalf("invalid xDS server type %q", x.config.Type)
	}
```

I didn't know that Contour has its own implementation of the xDS before you mentioned it now 😅 

So yeah you probably read below, ADS is required for this work. 

Now I am not sure if such a migration should happen in this design or not.

I am happy to work on making this migration in another work thread if you have an issue open or a design doc for the outcome.

Reply:
Yep that is the switch between the two xDS server implementations